pkg/machines/components/vcpuModal.jsx

```diff
@@ -0,0 +1,291 @@
+            state.sockets = divs[divs.length - 1];
+        }
+
+        // Likewise: Max value = Sockets * Cores * Threads. Cores = Max value / ( Threads * Sockets )
+        state.cores = (this.state.max / (state.sockets * state.threads));
+
+        this.setState(state);
+    }
+
+    onCoresChange (value) {
+        let state = { sockets: this.state.sockets, threads: this.state.threads };
+        state.cores = parseInt(value);
+
+        let divs = dividers(this.state.max / state.cores);
+
+        // If current sockets value is not in divisors array, then change it to max divisor
+        if (divs.indexOf(state.sockets) === -1) {
+            state.sockets = divs[divs.length - 1];
+        }
+
+        // Likewise: Max value = Sockets * Cores * Threads. Threads = Max value / ( Cores * Sockets )
+        state.threads = (this.state.max / (state.sockets * state.cores));
+        this.setState(state);
+    }
+    onOvirtChange (paramName) {
+        return (e) => {
+            let value = parseInt(e.target.value);
+            if (!value) {
+                value = 1;
+            }
+
+            let state = { sockets: this.state.sockets, threads: this.state.threads, cores: this.state.cores };
+            state[paramName] = value;
+            // Get values of topology and calculate product of multiplying
+            state.count = Object.values(state).reduce((accumulator, currentValue) => accumulator * currentValue, 1);
+
+            this.setState(state);
+        }
+    }
+
+    render () {
+        let caution = null;
+
+        if (this.props.isRunning && !this.props.isOvirt) {
+            caution = (
+                <tr>
+                    <td colSpan={2} className="machines-vcpu-caution">
+                        <Alert text={_("All changes will take effect only after reboot.")}/>
+                    </td>
+                </tr>);
+        }
+
+        let vcpuColumn = null;
+        let sctColumn = (<td>
+            <table className='form-table-ct vcpu-detail-modal-right'>
+                <InfoRecord descr={_("Sockets")} tooltip={_("Preferred number of sockets to expose to the guest.")} value={
+                    <Select id='socketsSelect' value={this.state.sockets.toString()} onChange={this.onSocketChange} items={dividers(this.state.max).map((t) => t.toString())} />
+                }/>
+                <InfoRecord descr={_("Cores per socket")} value={
+                    <Select id='coresSelect' value={this.state.cores.toString()} onChange={this.onCoresChange}  items={dividers(this.state.max).map((t) => t.toString())} />
+                }/>
+                <InfoRecord descr={_("Threads per cores")} value={
+                    <Select id='threadsSelect' value={this.state.threads.toString()} onChange={this.onThreadsChange} items={dividers(this.state.max).map((t) => t.toString())} />
+                }/>
+            </table>
+        </td>);
+
+        if (!this.props.isOvirt) {
```

Comment:
Likewise.
If provider needs UI change from the Libvirt base, then depending on complexity of the difference
- a specific component shall be implemented (see `pkg/ovirt/provider.es6:53-62` for examples)
- or predicates/value supplier functions in PROVIDER can be used (like `PROVIDR.isSomethingValid(vm, otherData)`)